### PR TITLE
Initial Setup

### PR DESCRIPTION
**Pull request by willmcclellan**
### Todo

* [x] Babel transpiling. I'm using `babel-tape` in the tests but the code needs to be compiled with a build step
* [x] Readme
* [x] Versioning
* [x] <del>Publish to npm</del> after PR us merged
* [x] Validation for relationship fieldtype
* [x] <del>Default values - switch should default to false, other fieldtypes might have optional user defined default</del> To be done in separate branch

The first bit of work on the plugin. The whole thing turned out to be a lot simpler than first anticipated, which is nice. The parts that make up the module are:

### Fieldtypes

Static representations of (UI-based) data structures that allow users to input content. Currently a fieldtype just has name (e.g. 'select') and a validation method

### Templates

Mongoose plugin that allows templating capabilities to be added to you schemas.

The idea is for this to be applied to a generic 'templates' collection where users can store templates, that are composed of 'fieldGroups'. A fieldGroup has the following properties:

* Label
* Description
* Fields (Array)
  * fieldtype (see above)
  * validation (available values, min, max etc)
  * options (to be used for display config on the client e.g.layout)
  
Once templates are created, they can be used to model user input on clients (apps, web apps etc). An example of a template would be some arbritary information that needs to be add to a report, for example. That might be a Date, Type of Report, and question and answer style input (using the switch fieldtype).

The template would be sent down to the client, which would render the relevant form structure using the JSON. Once the user has completed the form, it can be sent back up to the server to be stores in an entries colleciton.

### Entries

Entries is the second mongoose plugin. You can attach entries to any collection in mongoose, and you have the ability to store user generated content, based on templates.

The entry content uses the exact same schema as the template plugin, but utilises the `value` property on fields. This is where the value of the user input should be stored.

The duplication of the schema and template onto the entry is by design. It ensure a few things:

* There is no tight connection between a template and an entry. So if an entry is created based on a template and later that same template is updated, it has not affect on any entries previously submitted
* Because of this loose connection, an entry can still be updated after it has been created because it has all the relevant template structure stored in it's schema
* There is less relational overhead between templates and entries. The initial plan was to tightly link the template/entry so that when an entry was submitted it could be validated against the actual template. This meant another db call to fetch the required template and would have meant more db calls if the entry was ever to be edited.

So entries are based on templates, but have no relation to each other – apart from sharing the same schema.

### CODE REVIEW

File: lib/plugins/entry/index.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+ * Formulate Mongoose Entry Plugin
+ */
+
+import _ from 'lodash'
+import mongoose from 'mongoose'
+import fieldGroupSchema from '../field-group-schema'
+import fieldtypes from '../../fieldtypes'
+
+const Schema = mongoose.Schema
+const ObjectId = Schema.Types.ObjectId
```
Comment (rolandnsharp) [inline lib/plugins/entry/index.js]: mongoose, Schema, and ObjectId are unused.

Comment (willmcclellan) [inline lib/plugins/entry/index.js]: Good spot, the schema got moved to a separate file and these got left behind, will fix it up

**Comment by rolandnsharp** [comment]
:+1:

**Comment by rolandnsharp** [comment]
@willmcclellan , can we merge this in? I can add the default values if needed. So that I can work on creating validations.

**Comment by rolandnsharp** [comment]
I'll complete default values as a PR.

**Comment by sinakarimi** [comment]
I think this is good to merge for now, we can continue to make updates to it in the future. e.g. allowing multiple relationships to be added in a single record. :+1:

**Comment by willmcclellan** [comment]
Thanks guys, I'll merge this in

**Pull request merged**